cli/flox-rust-sdk/src/models/manifest/composite.rs

```diff
@@ -0,0 +1,276 @@
+#![allow(dead_code)] // TODO: Remove on first use.
+
+use std::iter::once;
+
+use flox_core::Version;
+use thiserror::Error;
+
+use super::typed::{
+    Manifest,
+    ManifestBuild,
+    ManifestContainerize,
+    ManifestHook,
+    ManifestInstall,
+    ManifestOptions,
+    ManifestProfile,
+    ManifestServices,
+    ManifestVariables,
+};
+
+#[derive(Error, Debug)]
+pub enum MergeError {}
+
+/// A collection of manifests to be merged with a `ManifestMergeStrategy`.
+struct CompositeManifest {
+    composer: Manifest,
+    deps: Vec<Manifest>,
+}
+
+impl CompositeManifest {
+    fn merge(&self, merger: impl ManifestMergeStrategy) -> Result<Manifest, MergeError> {
+        let Some(first_dep) = self.deps.first() else {
+            // No deps, just composer.
+            return Ok(self.composer.clone());
+        };
+
+        self.deps
+            .iter()
+            .skip(1) // First dep is used as initializer.
+            .chain(once(&self.composer)) // Composer goes last.
+            .try_fold(first_dep.clone(), |merged, next| {
+                merger.merge(merged, next.clone())
+            })
+    }
+}
+
+/// Strategy for merging two manifests which can then be applied iteratively for
+/// multiple manifests.
+trait ManifestMergeStrategy {
+    fn merge_version(manifest1: &Manifest, manifest2: &Manifest) -> Result<Version<1>, MergeError>;
+    fn merge_install(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestInstall, MergeError>;
+    fn merge_vars(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestVariables, MergeError>;
+    fn merge_hook(manifest1: &Manifest, manifest2: &Manifest) -> Result<ManifestHook, MergeError>;
+    fn merge_profile(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestProfile, MergeError>;
+    fn merge_options(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestOptions, MergeError>;
+    fn merge_services(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<ManifestServices, MergeError>;
+    fn merge_build(manifest1: &Manifest, manifest2: &Manifest)
+        -> Result<ManifestBuild, MergeError>;
+    fn merge_containerize(
+        manifest1: &Manifest,
+        manifest2: &Manifest,
+    ) -> Result<Option<ManifestContainerize>, MergeError>;
+    fn merge(&self, manifest1: Manifest, manifest2: Manifest) -> Result<Manifest, MergeError>;
+}
+
+/// Merges two manifests by applying `manifest2` on top of `manifest1` and
+/// overwriting any conflicts for keys within the top-level of each `Manifest`
+/// field, with the exception of `profile` and `hooks`.
+struct ShallowMerger;
```

Comment:
Suggestion: Maybe `OverwriteMerger` instead? I don't have intuition for what "shallow" merging would mean. I agree that "prune" and "leaf" don't provide a lot of intuition either.

Reply:
I feel like "shallow" and "deep" merging are better known terminology.

(setting aside that we're doing something completely different for hooks)

Reply:
Up to you, I'll just say that the only reason I know what `ShallowMerger` does is because I have out of band knowledge that we chose to merge a certain way.

Reply:
just a datapoint being that i also have more intuition for "shallow" vs "deep" merges.

That said, we dont consistently apply that in a principled way iff were appending profile scritpts.

Reply:
Alright, shallow it is

Reply:
I also prefer shallow and deep to leaf and prune even though that's what I was calling it